LG4 Rule Object Structure in Code

carlosmspk opened the pull request:
Data Class implementation of rules and scopes models (as well as a config model placeholder)

This PR depends on LG5 as that issue pertains to defining the structure of the rules file which implicitely changes the structure of the in-code model

usage (the two `CommitMessageSizeRule` are just an example, this scenario doesn't make any sense):
```python
ruleset = ScopedRuleSet(
    scope=BranchNameScope(action=ScopeAction.INCLUDE, name_like="develop"),
    rules=[
        CommitMessageSizeRule(max_size=100),
        CommitMessageSizeRule(max_size=50),
    ],
)
```

Which, if printed, would yield:
```python
ScopedRuleSet(scope=BranchNameScope(action=<ScopeAction.INCLUDE: 'include'>, name_like='develop'), rules=[CommitMessageSizeRule(type=RuleType.COMMIT_MESSAGE_SIZE, max_size=100), CommitMessageSizeRule(type=RuleType.COMMIT_MESSAGE_SIZE, max_size=50)])
```

carlosmspk committed:
LG4 Add Model package

carlosmspk committed:
LG4 Add base rule class and rule type enum

carlosmspk committed:
LG4 Add concrete rule class for commit message size

carlosmspk committed:
LG4 Add base scope class and scope action enum

carlosmspk committed:
LG4 Add concrete scope class for branch name

carlosmspk committed:
LG4 Add ScopedRuleSet class

carlosmspk committed:
LG4 Add placeholder Config class

dandepaiva commented on model/rules.py:
```diff
@@ -0,0 +1,31 @@
+    COMMIT_MESSAGE_SIZE = "CommitMessageSize"
+
+
+class Rule:
```
I think it would be good to have function/class documentation in these functions/classes from early.

carlosmspk replied:
Makes sense, should also come in handy for later steps, for documenting how the rules file fields work, and how to implement the rule checker

carlosmspk replied:
Added document strings on most classes and fields.
Also added README on models package for info on creating new Rules and Scopes as well as how they should work

carlosmspk committed:
LG4 Add vscode config file to gitignore

carlosmspk committed:
LG4 Rename Scope to ScopeCondition (and other scope-related classes accordingly)

carlosmspk committed:
LG4 Update list fields to be Iterable, becoming imutable

carlosmspk committed:
LG4 Add abstract package for util abstract class and fields features

carlosmspk committed:
LG4 Update ScopedRuleSet scope field to be made up of a list of Scope conditions

carlosmspk committed:
LG4 Add README to model package to document developer usage of contained modules

carlosmspk committed:
LG4 Update dataclasses to be imutable

carlosmspk committed:
LG4 Add docstrings to rule module classes

carlosmspk committed:
LG4 Rename rules to rule_set to better reflect ScopedRuleSet class name

carlosmspk committed:
LG4 Update abstract classes to use new abstrack package

carlosmspk committed:
LG4 Add docstrings to Config class

carlosmspk committed:
LG4 Refactor base Rule class repr method to not print redundant rule type field

carlosmspk committed:
LG4 Update CommitMessageSizeRule max_size field to be readonly property

carlosmspk committed:
LG4 Add docstrings to scope module classes

carlosmspk committed:
LG4 Rename ScopeCondition module file

carlosmspk committed:
LG4 Add module guards for runtime verification of abstract class instantiation

carlosmspk committed:
LG4 Remove redundant str method on CommitMessageSizeRule

carlosmspk committed:
LG4 Add exceptions module for custom exception types

carlosmspk committed:
LG4 Add docstrings to ScopedRuleSet dataclass and its fields

carlosmspk commented:
Further changes:
 - Added "util" package for abstract classes (Python doesn't support abstract classes all that well)
 - Added exceptions module in model package for custom exceptions made by us
 - Added .vscode to gitignore

dandepaiva commented on model/__init__.py:
is this empty file necessary?

carlosmspk replied:
Yes and no. In the good old days, Python required __init__.py files to define packages to be importable. This is no longer required since Python 3.3 (September 29, 2012), and I just do it as a force of habit. I'll remove them

dandepaiva commented on model/rules.py:
```diff
@@ -0,0 +1,45 @@
+
+
+class RuleType(Enum):
+    """Collection of rule types used to uniquely identify each subclass of Rule. The Enum's value is what should show up in the rule YAML file"""
```
Usage of the word "Collection" can lead to confusion.
By definition, this is not a collection.

dandepaiva commented on model/rules.py:
```diff
@@ -0,0 +1,45 @@
+    def __repr__(self) -> str:
+        """This is to allow all rules to be printable in a human-readable manner for debugging
+
+        Never call this method explicitely, this is Python's way of converting an object to string, which gets automatically
```
*explicitly
In my opinion, you can make a shorter version of this, since it will not be called explicitly, might as well just explain why it is here, "Automatically called after `print(my_rule)` (....)", might be enough, wdyt?

carlosmspk replied:
Changed to:
> Enum which identifies the rule's type. The Enum's value is what should show up in the rule YAML file

dandepaiva commented on model/ruleset.py:
```diff
@@ -0,0 +1,15 @@
+
+    scope: Iterable[ScopeCondition]
+    """A combination of scope conditions which defines the scope as a whole. The first scope condition that gets matched defines the outcome of whether rules should be evaluated or not, therefore scope condition order is relevant."""
+    rule_set: Iterable[Rule]
```
"ruleset" is acceptable as one word and will prevent possible confusion in the future.
Same for the class name.

carlosmspk replied:
Also changed in README

dandepaiva commented on abstract/__init__.py:
Is this empty file necessary?

dandepaiva commented on model/exceptions.py:
```diff
@@ -0,0 +1,7 @@
+class IllegalAbstractClassInitCall(Exception):
+    """Attempted to instantiate an abstract class"""
```
I would prefer: "Exception raised when an instantiation of an abstract class is attempted"

carlosmspk replied:
This is just Python's standard. Here are some examples of built-in exceptions and their docstrings:

- `TypeError`: _Inappropriate argument type._
- `KeyError`: _Mapping key not found._
- `NameError`: _Name not found globally._
- etc.
The rationale is that by having something extending Exception, it becomes obvious that it's an "Exception raised when..."
That being said, I can rename it to `AbstractClassInstantiatedError`?

dandepaiva replied:
Understood.
Yes, good idea, I think that would make it clearer

carlosmspk replied:
Renamed

carlosmspk committed:
LG4 Update RuleType enum docstring to no longer describe it as a collection

carlosmspk committed:
LG4 Update Rule __repr__ docstring to be more concise

carlosmspk committed:
LG4 Update all occurrences of "rule set" to "ruleset"

carlosmspk committed:
LG4 Remove deprecated __init__ package declaration files

dandepaiva commented on model/ruleset.py:
```diff
@@ -0,0 +1,15 @@
+
+@dataclass(frozen=True)
+class ScopedRuleset:
+    """Set of rules to be applied to matching set of scope conditions, depending on the matched scope action."""
```
Avoid using the word Set

carlosmspk replied:
Changed:
```
Set of rules to be applied to matching set of scope conditions...
```
to
```
Class with rules to be applied to matching scope conditions...
```

carlosmspk committed:
LG4 Rename module scopes_conditions to scope_conditions

carlosmspk committed:
LG4 Update ScopedRuleset docstring to not contain misleading "set" word

carlosmspk committed:
LG4 Rename IllegalAbstractClassInitCall to AbstractClassInstantiatedError to make its type more obvious

carlosmspk committed:
LG4 Fix ScopedRuleset docstring syntax

dandepaiva commented on model/README.md:
```diff
@@ -0,0 +1,84 @@
+
+#### Example - Create a new rule for maximum allowed diffs
+
+For a new rule which limits how many diffs a single commit can have, we could something like the following in [the rules file](./rules.py)
```
(...), we could ***have** something like (...)

carlosmspk replied:
Added

dandepaiva commented on model/README.md:
```diff
@@ -0,0 +1,84 @@
+# Scoped Rulesets
+
+Scoped Rulesets simply combine a set of scope conditions (i.e. a Scope) and associate a collection of rules (i.e. a Ruleset). Of particular importance:
+ - Scope conditions should be evaluated on first-match (order is relevant): check each scope condition, if it matches, use that scope condition's action immediately; if it does not match, carry on to check if the next scope condition matches, and so on... If not scope conditions match, by default, exclude ruleset
```
the phrasing here is confusing to me, 
after the ellipsis, is that "not" supposed to be there?

carlosmspk replied:
"not" should be "no": "... If no scope conditions match"

dandepaiva commented on model/README.md:
```diff
@@ -0,0 +1,84 @@
+Scoped Rulesets simply combine a set of scope conditions (i.e. a Scope) and associate a collection of rules (i.e. a Ruleset). Of particular importance:
+ - Scope conditions should be evaluated on first-match (order is relevant): check each scope condition, if it matches, use that scope condition's action immediately; if it does not match, carry on to check if the next scope condition matches, and so on... If not scope conditions match, by default, exclude ruleset
+
+ - Rules should be all be evaluated. The entire pipeline will fail if a single rule mismatches (rules define conditions for a valid git event). However, all rules should be validated, in order to collect a single result to report which rules failed
```
one "be", too many, at the start of the paragraph

carlosmspk replied:
Changed to "All rules should be evaluated", "rules should all be" sounded a bit awkward

carlosmspk committed:
LG4 Fix typos in README

carlosmspk committed:
LG4 Fix typo in README

carlosmspk committed:
LG4 Fix awkward sentence in README

dandepaiva commented:
Looks good to me.

carlosmspk committed:
LG4 Remove mindlessly copy pasted TODO

na-exaud commented on .gitignore:
```diff
@@ -158,3 +158,5 @@ cython_debug/
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+
+.vscode
```
Empty line at the end of the file or else github cries.

carlosmspk replied:
Added

na-exaud reviewed (approved):
Minor thing to resolve.

carlosmspk committed:
LG4 Add newline to .gitignore
